File: integration-tests-java/src/main/java/io/micronaut/jimmer/java/it/service/IBook.java
package io.micronaut.jimmer.java.it.service;

import io.micronaut.jimmer.java.it.entity.Book;
import java.math.BigDecimal;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import org.babyfish.jimmer.sql.ast.mutation.SimpleSaveResult;
import org.babyfish.jimmer.sql.fetcher.Fetcher;
import org.jetbrains.annotations.Nullable;

public interface IBook {

    Book findById(long id);

    Book findById(int id, Fetcher<Book> fetcher);

    List<Book> findByIds(List<Integer> ids);

    SimpleSaveResult<Book> save(Book book);

    Map<Long, BigDecimal> findAvgPriceGroupByStoreId(Collection<Long> storeIds);

    List<Book> findBooksByName(@Nullable String name, @Nullable Fetcher<Book> fetcher);

    List<Book> findBooksByName(@Nullable String name);

    void update();

    List<Book> manyToMany();

    void updateOneToMany();

    void saveManyToMany();
}

File: integration-tests-kotlin/src/main/kotlin/io/micronaut/jimmer/kotlin/it/service/IBook.kt
package io.micronaut.jimmer.kotlin.it.service

import io.micronaut.jimmer.kotlin.it.entity.Book
import org.babyfish.jimmer.sql.fetcher.Fetcher
import org.babyfish.jimmer.sql.kt.ast.mutation.KSimpleSaveResult
import java.math.BigDecimal

interface IBook {
    fun findById(id: Long): Book?

    fun findById(
        id: Int,
        fetcher: Fetcher<Book>,
    ): Book?

    fun findByIds(ids: List<Int>): List<Book>?

    fun save(book: Book): KSimpleSaveResult<Book>?

    fun findAvgPriceGroupByStoreId(storeIds: Collection<Long>): Map<Long, BigDecimal>

    fun findBooksByName(
        name: String,
        fetcher: Fetcher<Book>,
    ): List<Book>?

    fun findBooksByName(name: String): List<Book>?

    fun update()

    fun manyToMany(): List<Book>?

    fun updateOneToMany()

    fun saveManyToMany()
}
